src/ray/core_worker/core_worker.cc

```diff
@@ -405,51 +404,31 @@ Status CoreWorker::Get(const std::vector<ObjectID> &ids, const int64_t timeout_m
   bool got_exception = false;
   absl::flat_hash_map<ObjectID, std::shared_ptr<RayObject>> result_map;
   auto start_time = current_time_ms();
-  RAY_RETURN_NOT_OK(plasma_store_provider_->Get(
-      plasma_object_ids, timeout_ms, worker_context_, &result_map, &got_exception));
 
-  if (!got_exception) {
-    int64_t local_timeout_ms = timeout_ms;
-    if (timeout_ms >= 0) {
-      local_timeout_ms = std::max(static_cast<int64_t>(0),
-                                  timeout_ms - (current_time_ms() - start_time));
-    }
-    RAY_RETURN_NOT_OK(memory_store_->Get(memory_object_ids, local_timeout_ms,
-                                         worker_context_, &result_map, &got_exception));
+  if (!memory_object_ids.empty()) {
+    RAY_RETURN_NOT_OK(memory_store_->Get(memory_object_ids, timeout_ms, worker_context_,
+                                         &result_map, &got_exception));
   }
 
   if (!got_exception) {
     // If any of the objects have been promoted to plasma, then we retry their
     // gets at the provider plasma. Once we get the objects from plasma, we flip
     // the transport type again and return them for the original direct call ids.
-    absl::flat_hash_set<ObjectID> promoted_plasma_ids;
     for (const auto &pair : result_map) {
       if (pair.second->IsInPlasmaError()) {
-        RAY_LOG(DEBUG) << pair.first << " in plasma, doing fetch-and-get";
-        promoted_plasma_ids.insert(
-            pair.first.WithTransportType(TaskTransportType::RAYLET));
+        RAY_LOG(INFO) << pair.first << " in plasma, doing fetch-and-get";
+        plasma_object_ids.insert(pair.first);
       }
     }
-    if (!promoted_plasma_ids.empty()) {
-      int64_t local_timeout_ms = timeout_ms;
-      if (timeout_ms >= 0) {
-        local_timeout_ms = std::max(static_cast<int64_t>(0),
-                                    timeout_ms - (current_time_ms() - start_time));
-      }
-      RAY_LOG(DEBUG) << "Plasma GET timeout " << local_timeout_ms;
-      RAY_RETURN_NOT_OK(plasma_store_provider_->Get(promoted_plasma_ids, local_timeout_ms,
-                                                    worker_context_, &result_map,
-                                                    &got_exception));
-      for (const auto &id : promoted_plasma_ids) {
-        auto it = result_map.find(id);
-        if (it == result_map.end()) {
-          result_map.erase(id.WithTransportType(TaskTransportType::DIRECT));
-        } else {
-          result_map[id.WithTransportType(TaskTransportType::DIRECT)] = it->second;
-        }
-        result_map.erase(id);
-      }
+    int64_t local_timeout_ms = timeout_ms;
+    if (timeout_ms >= 0) {
+      local_timeout_ms = std::max(static_cast<int64_t>(0),
+                                  timeout_ms - (current_time_ms() - start_time));
     }
+    RAY_LOG(DEBUG) << "Plasma GET timeout " << local_timeout_ms;
+    RAY_RETURN_NOT_OK(plasma_store_provider_->Get(plasma_object_ids, local_timeout_ms,
```

Comment:
Nice, this would ensure we will not call plasma's Get for objects that do not exist.